### ISSUE TITLE
Cannot get uid for user <removed>

### ISSUE THREAD

**Issue by PunKeel** [AnoopAlias/nDeploy#62]
Hello,

I "still" (?) have the issue where deleted accounts are still present in php-fpm's config : 
```
root@web-01:~# grep rusth /opt/nDeploy/watcher.log
[03-Feb-2016 13:07:03] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 13:07:07] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 13:07:11] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:02:49] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:02:51] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:02:55] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:02:56] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:00] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:01] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:05] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:07] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:11] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:13] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'
[03-Feb-2016 17:03:17] ERROR: [pool rusthisp] cannot get uid for user 'rusthisp'


root@web-01:~# grep rusth /var/cpanel/accounting.log
Sat Oct 24 20:52:49 2015:CREATE:root:root:domain.es:5.135.103.187:rusthisp
Wed Feb  3 00:00:34 2016:REMOVE:root:root:domain.es:rusthisp
```
And so some domains (??, why not All ?) are giving 503 Errors

When running the nDeploy_watcher restart command, everything is fine again.

I'm using the last version (updated daily) .. Do you have any clue/need more information ?

**Comment by AnoopAlias** [comment]
@PunKeel 

Line #51 to #55 of https://github.com/AnoopAlias/nDeploy/blob/master/scripts/accountremove_hook_pre.py

should remove the pool conf file and reload the master

Check for 

nDeploy:remove:rusthisp

in /usr/local/cpanel/logs/error_log 

Perhaps the log entries just before this would give you a clue as to why the file was not removed.

nDeploy_watcher restart - You meant ndeploy_backends restart right?. backend restart will remove all stale php-fpm pool confs without a valid user in passwd file . You have to know why the file wasnt removed by the WHM hook in first place to get to the bottom of this

**Comment by PunKeel** [comment]
`grep 'nDeploy:remove:rusthisp' /usr/local/cpanel/logs/error_log` : no output

(Yes, I meant backend, sorry)
>backend restart will remove all stale php-fpm pool confs without a valid user in passwd file
Good to know!

**Comment by AnoopAlias** [comment]
grep rusthisp /usr/local/cpanel/logs/error_log | grep -i nDeploy

**Comment by PunKeel** [comment]
Nothing either

**Comment by AnoopAlias** [comment]
Can you check the hook is active. It can be checked from WHM hooks section

accountremove_hook_pre.py - this should be there

**Comment by PunKeel** [comment]
I don't understand it at all, so I'll just screenshot ... x)

![](http://j.ungeek.eu/hlhXJY5Cl)

**Comment by AnoopAlias** [comment]
yes hooks there . Something prevented the hook from running or it run with errors.

The cpanel error log is where you will get the details of what happened when the account was terminated.

**Comment by PunKeel** [comment]
Nothing with `grep rusthisp /usr/local/cpanel/logs/error_log|grep -v backup` either ... What's wrong with you, cPanel?

**Comment by AnoopAlias** [comment]
@PunKeel  - Can you try an upgrade to nDeploy-2.0-26
https://support.sysally.net/news/59

The old code contained a php-fpm master reload for each domain and its list of subdomains which was not required .The master is reloaded much fewer ( once for every user account add and/or delete ) and it should be less error prone now being a lot simplified.

Let me know if you see any more issues after the upgrade

**Comment by PunKeel** [comment]
Hello,

I've upgraded and it happened again (same symptoms)... :(

**Comment by AnoopAlias** [comment]
Create a dummy account  switch to php backend and terminate it ..tail the cpanel error_log when terminating .See if the pool file is getting removed or from the error log some errors are there when the hook gets executed

**Comment by AnoopAlias** [comment]
FYI ..the last nDeploy upgrade had a small bug in pool restart which is fixed in newer release .But that shouldn't affect pool file removal. In any case do an upgrade before you test this .

**Comment by PunKeel** [comment]
When creating an account manually I have this error:
```
Conf:Gen pun
Traceback (most recent call last):
  File "/opt/nDeploy/scripts/generate_config.py", line 558, in <module>
    nginx_confgen(cpaneluser, main_domain)  # Generate conf for main domain
  File "/opt/nDeploy/scripts/generate_config.py", line 488, in nginx_confgen
    nginx_confgen_profilegen(user_name, domain_sname, cpanel_ipv4, document_root, 0, domain_home, *domain_aname_list)
  File "/opt/nDeploy/scripts/generate_config.py", line 168, in nginx_confgen_profilegen
    with open("/var/cpanel/users/" + user_name) as users_file:
IOError: [Errno 2] No such file or directory: '/var/cpanel/users/pun'
User::Data::Modify pun
```

When trying to access the "nginx/ndeploy" section, it tells me `ERROR : domain-data file i/o error`.

**Comment by punkeel** [comment]
When deleting an existing account named "punkeel": (err log)

```
[2016-02-08 21:58:29 +0100] info [whostmgr] End STDERR from hook
[2016-02-08 21:58:31 +0100] warn [updateuserdomains] domain conflict: /var/cpanel/users/system contains a domain already owned by meterre1
proftpd: no process found
Cannot find disk data
[2016-02-08 21:58:31 +0100] info [whostmgr] killacct::Notify Notification => web@ni-hos***** via EMAIL [eventimportance => Low (3)]
```

watcher log:

```
Conf:Del /etc/nginx/sites-enabled/cache.json.* /opt/nDeploy/domain-data/cache.json
nginx: [warn] conflicting server name "www.culture-xxx.fr" on xxx.171.70:80, ignored
nginx: [warn] conflicting server name "culture-xxx.fr" on xxx.171.70:80, ignored
Conf:Del /etc/nginx/sites-enabled/xxx.fr.* /opt/nDeploy/domain-data/xxx.fr
nginx: [warn] conflicting server name "www.culture-xxx.fr" on xxx.171.70:80, ignored
nginx: [warn] conflicting server name "culture-ixxx.fr" on xxx.171.70:80, ignored
Conf:Gen punkeel
```

**Comment by PunKeel** [comment]
I think I fixed it ... had a domain in `/var/cpanel/users/system` as said in the logs, and it somehow broke everything.

I'm not yet sure at 1000 percent, will tell you if it fails again.

Thanks a lot for your time, @AnoopAlias !

**Comment by AnoopAlias** [comment]
cool. Do let me know . I like to keep the issue's tab at 0  (old school helpdesk support guy syndrome ;) )

**Comment by PunKeel** [comment]
Looks like it's not fixed :/

```
[10-Feb-2016 22:16:35] ERROR: [pool nameorp] cannot get uid for user 'nameorp'
[10-Feb-2016 22:16:35] ERROR: FPM initialization failed
```

```
root@ni-web-01:~# exportgrep -A 5 -B 5 nameorp /opt/nDeploy/watcher.log
Traceback (most recent call last):
  File "/opt/nDeploy/scripts/generate_config.py", line 558, in <module>
    nginx_confgen(cpaneluser, main_domain)  # Generate conf for main domain
  File "/opt/nDeploy/scripts/generate_config.py", line 488, in nginx_confgen
    nginx_confgen_profilegen(user_name, domain_sname, cpanel_ipv4, document_root, 0, domain_home, *domain_aname_list)
  File "/opt/nDeploy/scripts/generate_config.py", line 168, in nginx_confgen_profilegen
    with open("/var/cpanel/users/" + user_name) as users_file:
IOError: [Errno 2] No such file or directory: '/var/cpanel/users/nameorp'
User::Data::Modify nameorp
Conf:Gen nameorp
Conf:Del /etc/nginx/sites-enabled/nameo-.eu.* /opt/nDeploy/domain-data/nameo-rp.eu
```

**Comment by AnoopAlias** [comment]
@PunKeel -I think there is something wrong at how cPanel is setup in your system. 

/opt/nDeploy/scripts/generate_config.py checks first if the user is present in the system(passwd file) and if yes exits

But in your case nameorp is present in password ;but the file /var/cpanel/users/nameorp is missing 

Try /scripts/updateuserdomains
and do a forced upcp to the latest cpanel RELEASE .

**Comment by PunKeel** [comment]
Command ran without output, and upcp is up to date

**Comment by AnoopAlias** [comment]
@PunKeel  - Can you add the following PubKey in the server and send the servers main IP and port to 
support@sysally.net .Mention this ticket .

https://gist.github.com/AnoopAlias/d4a991630ceefd419482

**Comment by AnoopAlias** [comment]
@PunKeel - I checked this up and here's whats happening 
1. When the account is created the files in /var/cpanel/userdata are created trigerring the config generation;but the script errors out because /var/cpanel/users/USERNAME gets created after a small delay .So the config generation process does not complete.

2. The account_remove_hook.py does get run and remove the pool file .But the /var/cpanel/userdata files get removed only after a slight delay and this cause confGeneration to trigger again and recreate the pool file. This was confirmed by checking the pool files inode num . The stray pool file has a different inode number than the original.

To summarize both are bugs and both should have been taken care of in nDeploy . But this has been left unnoticed as small delay mentioned is only something I observe in your server. 

I have tagged this as bug and hopefully will fix this today.

Will keep you posted.
